jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferOutputStream2.java

```diff
@@ -0,0 +1,144 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.util.BufferUtil;
+
+/**
+ * This class implements an output stream in which the data is written into a list of ByteBuffer,
+ * the buffer list automatically grows as data is written to it, the buffers are taken from the
+ * supplied {@link ByteBufferPool} or freshly allocated if one is not supplied.
+ *
+ * Designed to mimic {@link java.io.ByteArrayOutputStream} but with better memory usage, and less copying.
+ */
+public class ByteBufferOutputStream2 extends OutputStream
+{
+    private final ByteBufferAccumulator _accumulator;
+    private final ByteBufferPool _bufferPool;
+    private ByteBuffer _combinedByteBuffer;
```

Comment:
Is there really a need to remember the combined ByteBuffer?

Reply:
We need to remember it so that we release it when `close()` is called.

Reply:
I think it is wrong for this class to manage the releasing of the combined buffer.  Isn't that mostly passed along a chain and will be released when processing is complete.    I still lean towards `takeBuffer` semantics.

Reply:
I think this behaviour can be implemented in `ByteBufferAccumulator` with 2 methods:
 - `toBuffer` will create a single buffer of all the contents, it will release any buffers in the list and the list will end up containing just the new single buffer.  Any subsequent calls to `toBuffer` will get the same buffer back. Any calls to add more content will just allocate another buffer in the list.  Buffer will be release with `close`
 - `takeBuffer` will create a single buffer of all the contents, it will release any buffers in the list and clear it.   It is the callers responsibility to release the returned buffer.